src/Common/src/TypeSystem/Common/MetadataType.cs

```diff
@@ -59,6 +59,26 @@ public override bool HasStaticConstructor
         /// Returns true if the type has given custom attribute.
         /// </summary>
         public abstract bool HasCustomAttribute(string attributeNamespace, string attributeName);
+
+        /// <summary>
+        /// Adapt current to the <paramref name="typeInstantiation"/> context. For example, if you have C <T> and a
+        /// context [int], it will yield C <int>.
+        /// </summary>
+        /// <param name="typeInstantiation">Context</param>
+        /// <param name="methodInstantiation">Not used for types.</param>
+        /// <returns></returns>
+        public override TypeDesc InstantiateSignature(Instantiation typeInstantiation, Instantiation methodInstantiation)
```

Comment:
This should not be override of `InstantiateSignature`. If it is needed (after incorporating other feedback) - it should be a helper method (probably on TypeSystemHelpers) that does something like:
``
return type.HasInstantiation ? type.MakeInstantiatedType(type.GetTypeDefinition(), typeInstantiation) : type;
```